Source/pack.cpp

```diff
@@ -68,34 +87,108 @@ void PackNetItem(const Item &item, ItemNetPack &packedItem)
+
+	if (isPregenItem && hasMultipleFlags(iCreateInfo))
+		return false;
+	if (isGroundItem && (iCreateInfo & ~CF_USEFUL) != 0)
+		return false;
+	if (isTownItem && hasMultipleFlags(iCreateInfo))
+		return false;
+	return true;
 }
 
-template <typename T>
-void LogFailedJoinAttempt(const char *condition, const char *name, T value)
+bool IsTownItemValid(uint16_t iCreateInfo)
 {
-	LogDebug("Remote player validation failed: ValidateField({}: {}, {})", name, value, condition);
+	const uint8_t level = iCreateInfo & CF_LEVEL;
+	const bool isBoyItem = (iCreateInfo & CF_BOY) != 0;
+
+	if (isBoyItem && level <= MaxCharacterLevel)
+		return true;
+
+	return level <= 30;
 }
 
-template <typename T1, typename T2>
-void LogFailedJoinAttempt(const char *condition, const char *name1, T1 value1, const char *name2, T2 value2)
+bool IsUniqueMonsterItemValid(uint16_t iCreateInfo, uint32_t dwBuff)
 {
-	LogDebug("Remote player validation failed: ValidateFields({}: {}, {}: {}, {})", name1, value1, name2, value2, condition);
+	const uint8_t level = iCreateInfo & CF_LEVEL;
+	const bool isHellfireItem = (dwBuff & CF_HELLFIRE) != 0;
+
+	for (int i = 0; UniqueMonstersData[i].mName != nullptr; i++) {
+		const auto &uniqueMonsterData = UniqueMonstersData[i];
+		const auto &uniqueMonsterLevel = static_cast<uint8_t>(MonstersData[uniqueMonsterData.mtype].level);
+
+		if (!isHellfireItem && IsAnyOf(uniqueMonsterData.mtype, MT_HORKDMN, MT_DEFILER, MT_NAKRUL)) {
+			// These monsters don't appear in Diablo
+			continue;
+		}
+
+		if (level == uniqueMonsterLevel) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+bool IsDungeonItemValid(uint16_t iCreateInfo, uint32_t dwBuff)
+{
+	const uint8_t level = iCreateInfo & CF_LEVEL;
+	const bool isHellfireItem = (dwBuff & CF_HELLFIRE) != 0;
+
+	for (int i = 0; MonstersData[i].name != nullptr; i++) {
+		const auto &monsterData = MonstersData[i];
+		auto monsterLevel = static_cast<uint8_t>(monsterData.level);
+
+		if (IsAnyOf(i, MT_HORKDMN, MT_DEFILER, MT_NAKRUL)) {
+			continue;
+		}
+
+		if (i == MT_DIABLO && !isHellfireItem) {
+			monsterLevel -= 15;
+		}
+
+		if (level == monsterLevel) {
+			return true;
+		}
+	}
+
+	return level <= 30;
+}
+
+bool UnPackNetItem(const Player &player, const ItemNetPack &packedItem, Item &item)
+{
+	item = {};
+	_item_indexes idx = static_cast<_item_indexes>(SDL_SwapLE16(packedItem.def.wIndx));
+	if (idx < 0 || idx > IDI_LAST)
+		return false;
+	if (idx == IDI_EAR) {
+		RecreateEar(item, SDL_SwapLE16(packedItem.ear.wCI), SDL_SwapLE32(packedItem.ear.dwSeed), packedItem.ear.bCursval, packedItem.ear.heroname);
+		return true;
+	}
+
+	uint16_t creationFlags = SDL_SwapLE16(packedItem.item.wCI);
+	uint32_t dwBuff = SDL_SwapLE16(packedItem.item.dwBuff);
+	ValidateField(creationFlags, IsCreationFlagComboValid(creationFlags));
+	if ((creationFlags & CF_TOWN) != 0)
+		ValidateField(creationFlags, IsTownItemValid(creationFlags));
+	else if ((creationFlags & CF_UPER15) != 0 && (creationFlags & CF_USEFUL) != CF_USEFUL)

```

Comment:
This is the same as checking that CF_UPER15 is set but CF_UPER1 is not. Which `IsCreationFlagComboValid()` already asserted is not the case.

Reply:
No, because `CF_USEFUL = CF_UPER15 | CF_UPER1`, indicating it's a ground item. Ground items can have both these flags set so long as all other flags are unset.

Reply:
Yeah you are checking that `CF_UPER15` is set, and that the `CF_UPER1 | CF_UPER15` isn't. Meaning you can remove `CF_UPER15` from the right side.

Reply:
Sorry, I think I came across more harshly than I intended. What I meant to say is that `IsCreationFlagComboValid()` didn't already assert that the item is not `CF_USEFUL`. I guess you already knew that `CF_USEFUL = CF_UPER15 | CF_UPER1`, and you are correct that the expression can be simplified.